solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpSolrClientJdkImpl.java

```diff
@@ -0,0 +1,443 @@
+      contentType = contentWriter.getContentType();
+    }
+    reqb.header("Content-Type", contentType);
+
+    if (isMultipart(streams)) {
+      throw new UnsupportedOperationException("This client does not support multipart.");
+    }
+
+    if (contentWriter != null) {
+      // TODO:  There is likely a more memory-efficient way to do this!
+      ByteArrayOutputStream baos = new ByteArrayOutputStream();
+      contentWriter.write(baos);
+      byte[] bytes = baos.toByteArray();
+      return new PreparePostPutRequestResult(
+          queryParams, HttpRequest.BodyPublishers.ofByteArray(bytes));
+    } else if (streams != null && streams.size() == 1) {
+      ContentStream contentStream = streams.iterator().next();
+      InputStream is = contentStream.getStream();
+      return new PreparePostPutRequestResult(
+          queryParams, HttpRequest.BodyPublishers.ofInputStream(() -> is));
+    } else if (queryParams != null && urlParamNames != null) {
+      ModifiableSolrParams requestParams = queryParams;
+      queryParams = calculateQueryParams(urlParamNames, requestParams);
+      queryParams.add(calculateQueryParams(solrRequest.getQueryParams(), requestParams));
+      return new PreparePostPutRequestResult(
+          queryParams, HttpRequest.BodyPublishers.ofString(requestParams.toString()));
+    } else {
+      return new PreparePostPutRequestResult(queryParams, HttpRequest.BodyPublishers.noBody());
+    }
+  }
+
+  private void decorateRequest(HttpRequest.Builder reqb, SolrRequest<?> solrRequest) {
+    if (requestTimeoutMillis > 0) {
+      reqb.timeout(Duration.of(requestTimeoutMillis, ChronoUnit.MILLIS));
+    } else {
+      reqb.timeout(Duration.of(idleTimeoutMillis, ChronoUnit.MILLIS));
+    }
+    reqb.header("User-Agent", USER_AGENT);
+    setBasicAuthHeader(solrRequest, reqb);
+    Map<String, String> headers = solrRequest.getHeaders();
+    if (headers != null) {
+      for (Map.Entry<String, String> entry : headers.entrySet()) {
+        reqb.header(entry.getKey(), entry.getValue());
+      }
+    }
+  }
+
+  private void setBasicAuthHeader(SolrRequest<?> solrRequest, HttpRequest.Builder reqb) {
+    if (solrRequest.getBasicAuthUser() != null && solrRequest.getBasicAuthPassword() != null) {
+      String encoded =
+          basicAuthCredentialsToAuthorizationString(
+              solrRequest.getBasicAuthUser(), solrRequest.getBasicAuthPassword());
+      reqb.header("Authorization", encoded);
+    } else if (basicAuthAuthorizationStr != null) {
+      reqb.header("Authorization", basicAuthAuthorizationStr);
+    }
+  }
+
+  private static final Pattern MIME_TYPE_PATTERN = Pattern.compile("^(.*?)(?:;| |$)");
+
+  private String contentTypeToMimeType(String contentType) {
+    Matcher mimeTypeMatcher = MIME_TYPE_PATTERN.matcher(contentType);
+    return mimeTypeMatcher.find() ? mimeTypeMatcher.group(1) : null;
+  }
+
+  private static final Pattern CHARSET_PATTERN = Pattern.compile("(?i)^.*charset=(.*)?(?:;| |$)");
```

Comment:
Prior to "charset" we can expect a space, I think.  And the dot-star you are matching should not be whitespace or a semicolon; then you could eliminate the non-capturing group that follows.

Reply:
I simplified it and added a unit test.